volttron/platform/agent/base_historian.py

```diff
@@ -1518,41 +1520,54 @@ def remove_successfully_published(self, successful_publishes,
         If None is found in `successful_publishes` we assume that everything
         was published.
 
-        :param successful_publishes: List of records that was published.
+        :param successful_publishes: Set of records that was published.
         :param submit_size: Number of things requested from previous call to
                             :py:meth:`get_outstanding_to_publish`
 
-        :type successful_publishes: list
+        :type successful_publishes: set
         :type submit_size: int
 
         """
 
         #_log.debug("Cleaning up successfully published values.")
         c = self._connection.cursor()
-
-        if None in successful_publishes:
-            c.execute('''DELETE FROM outstanding
-                        WHERE ROWID IN
-                        (SELECT ROWID FROM outstanding
-                          ORDER BY ts LIMIT ?)''', (submit_size,))
-            if self._record_count < c.rowcount:
-                self._record_count = 0
+        try:
+            if None in successful_publishes:
+                if not self._dupe_ids:
+                    c.execute('''DELETE FROM outstanding
+                                WHERE ROWID IN
+                                (SELECT ROWID FROM outstanding ORDER BY ts LIMIT ?)''', (submit_size,))
+                    if self._record_count < c.rowcount:
+                        self._record_count = 0
+                    else:
+                        self._record_count -= c.rowcount
+                else:
+                    # this is a rare case, so adding logs to help track and debug when this case does occur
+                    _log.debug(f"Duplicates detected during successful publishes. Keeping the following duplicate records in cache: {self._dupe_ids}")
+                    _log.debug(f"Deleting the following unique records: {self._unique_ids}")
+                    c.executemany('''DELETE FROM outstanding 
+                                              WHERE id = ?''',
+                                  ((_id,) for _id in self._unique_ids))
+                    self._record_count -= len(self._unique_ids)
             else:
-                self._record_count -= c.rowcount
-        else:
-            temp = list(successful_publishes)
-            temp.sort()
-            c.executemany('''DELETE FROM outstanding
-                            WHERE id = ?''',
-                          ((_id,) for _id in
-                           successful_publishes))
-            self._record_count -= len(temp)
+                temp = list(successful_publishes)
```

Comment:
there could be duplicate records even in the case were not all records in the batch got into db.

Reply:
Can you elaborate on this particular case? How would this happen in the historian execution flow?